### PR TITLE
ENGESC-7596 copy image tar into bucket based on the stack location and with rewrite SDK call

### PR DESCRIPTION
**Pull request by biharitomi**
**Main changes:**
 - enhanced logging
 - polling of rewrite status, Dyngr poller as new dependency
 - pre-check of bucket existence
 - error handling for existing bucket name, image and rewrite call read timeout
 - set the same location as the cluster for the created bucket
 - increased timeout for storage related SDK calls

### CODE REVIEW

**Comment by biharitomi** [comment]
retest this please

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java [hunk 1]
```diff
@@ -146,4 +139,111 @@ private boolean containsSpecificImage(ImageList imageList, String imageUrl) {
...
+        try {
+            storage.buckets().get(bucketName).execute();
+            existingBucket = true;
+        } catch (GoogleJsonResponseException ex) {
+            existingBucket = false;
+            if (ex.getStatusCode() == NOT_FOUND) {
+                LOGGER.warn("Bucket {} does not exist on provider side so we will create it: {}",
+                        bucketName, ex.getMessage());
+            } else {
+                LOGGER.warn("We were not able to get the bucket from Google side with name {} with exception {}. "
```
Comment (bergerdenes) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:154]: in this case are we going to create the bucket?

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java [hunk 2]
```diff
@@ -146,4 +139,111 @@ private boolean containsSpecificImage(ImageList imageList, String imageUrl) {
...
+            }
+        } catch (IOException ioException) {
+            LOGGER.error(errorMessage, ioException);
+            throw new CloudConnectorException(errorMessage, ioException);
+        }
+        return bucket;
+    }
+
+    private void copyImageTarToBucket(
+            final String sourceBucket,
```
Comment (bergerdenes) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:197]: we are not using `final` modifier on parameters

**Review by doktoric — APPROVED** [review]

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java [hunk 3]
```diff
@@ -146,4 +139,111 @@ private boolean containsSpecificImage(ImageList imageList, String imageUrl) {
...
+            LOGGER.warn("Unexpected error occurred when we tried to get bucket {} from Google side: {}",
+                    bucketName, e.getMessage());
+            existingBucket = true;
+        }
+        return existingBucket;
+    }
+
+    private Bucket createBucketIfNotExists(String projectId, AuthenticatedContext authenticatedContext, CloudContext cloudContext, Storage storage) {
+        String gcpLocation = authenticatedContext.getCloudContext().getLocation().getRegion().getRegionName();
+        String bucketName = String.format("%s-%s-%d", projectId, cloudContext.getName(), cloudContext.getId());
```
Comment (doktoric) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:168]: the problem will be the same here cloudContext.getId() is dynamic

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java [hunk 4]
```diff
@@ -146,4 +139,111 @@ private boolean containsSpecificImage(ImageList imageList, String imageUrl) {
...
+                    + "Please check Google Cloud console because probably the image should be ready.");
+        } catch (PollerException exception) {
+            String message = String.format("Polling failed for image copy to bucket '%s' with key '%s'", destBucket, destKey);
+            LOGGER.error(message, exception);
+            throw new CloudConnectorException("Image copy failed because: " + exception.getMessage(), exception);
+        } catch (Exception e) {
+            String message = String.format("Polling could not started to bucket '%s' with key '%s'", destBucket, destKey);
+            LOGGER.error(message, e);
+            throw new CloudConnectorException("Copying the image could not started, "
+                    + "please check that you already giving access to CDP for storage API.", e);
```
Comment (bergerdenes) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:226]: do we call Cloudbreak as CDP in the context of rc-2.9?

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java [hunk 5]
```diff
@@ -146,4 +139,111 @@ private boolean containsSpecificImage(ImageList imageList, String imageUrl) {
...
+        } catch (PollerStoppedException pollerStoppedException) {
+            LOGGER.error("Poller stopped for image copy: ", pollerStoppedException);
+            throw new CloudConnectorException("Image copy failed because the copy take too long time. "
+                    + "Please check Google Cloud console because probably the image should be ready.");
+        } catch (PollerException exception) {
+            String message = String.format("Polling failed for image copy to bucket '%s' with key '%s'", destBucket, destKey);
+            LOGGER.error(message, exception);
+            throw new CloudConnectorException("Image copy failed because: " + exception.getMessage(), exception);
+        } catch (Exception e) {
+            String message = String.format("Polling could not started to bucket '%s' with key '%s'", destBucket, destKey);
```
Comment (bergerdenes) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:223]: grammar: could not be started

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java [hunk 6]
```diff
@@ -146,4 +139,111 @@ private boolean containsSpecificImage(ImageList imageList, String imageUrl) {
...
+            LOGGER.error(message, e);
+            throw new CloudConnectorException("Copying the image could not started, "
```
Comment (bergerdenes) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:225]: grammar: could not be started

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java [hunk 7]
```diff
@@ -146,4 +139,111 @@ private boolean containsSpecificImage(ImageList imageList, String imageUrl) {
...
+                    .waitPeriodly(SLEEP_TIME_IN_SECONDS, TimeUnit.SECONDS)
+                    .run(new GcpImageAttemptMaker(rewriteResponse.getRewriteToken(), sourceBucket, sourceKey, destBucket, destKey, storage));
+            LOGGER.info("Image copy has been finished successfully for {}/{}.", destBucket, destKey);
+        } catch (UserBreakException userBreakException) {
+            LOGGER.error("Polling exited before timeout. Cause ", userBreakException);
+            throw new CloudConnectorException("The image copy failed because the one of the user in your "
+                    + "organization stopped the copy process.");
+        } catch (PollerStoppedException pollerStoppedException) {
+            LOGGER.error("Poller stopped for image copy: ", pollerStoppedException);
+            throw new CloudConnectorException("Image copy failed because the copy take too long time. "
```
Comment (bergerdenes) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:216]: grammar: has taken

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java [hunk 8]
```diff
@@ -146,4 +139,111 @@ private boolean containsSpecificImage(ImageList imageList, String imageUrl) {
...
+            Storage.Objects.Rewrite rewrite = storage.objects().rewrite(sourceBucket, sourceKey, destBucket, destKey, new StorageObject());
+            RewriteResponse rewriteResponse = rewrite.execute();
+            Polling.stopAfterAttempt(MAX_ATTEMPT_COUNT)
+                    .stopIfException(true)
+                    .waitPeriodly(SLEEP_TIME_IN_SECONDS, TimeUnit.SECONDS)
+                    .run(new GcpImageAttemptMaker(rewriteResponse.getRewriteToken(), sourceBucket, sourceKey, destBucket, destKey, storage));
+            LOGGER.info("Image copy has been finished successfully for {}/{}.", destBucket, destKey);
+        } catch (UserBreakException userBreakException) {
+            LOGGER.error("Polling exited before timeout. Cause ", userBreakException);
+            throw new CloudConnectorException("The image copy failed because the one of the user in your "
```
Comment (bergerdenes) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:212]: grammar: users

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java (hunk 4 above)
Comment (bergerdenes) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:226]: grammar: you are already giving

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java [hunk 9]
```diff
@@ -146,4 +139,111 @@ private boolean containsSpecificImage(ImageList imageList, String imageUrl) {
...
+            gcpApiImage.setRawDisk(rawDisk);
+            Insert ins = compute.images().insert(projectId, gcpApiImage);
+            ins.execute();
+        } catch (GoogleJsonResponseException googleJsonResponseException) {
+            if (googleJsonResponseException.getStatusCode() == HttpStatus.SC_CONFLICT) {
+                LOGGER.info("Conflict, the image already exists with name '{}', continuing without error", imageName);
+            } else {
+                String message = String.format("Failed to create image with name '%s' from tar: '%s' in bucket: '%s'", imageName, tarName, bucket.getName());
+                LOGGER.error(message, googleJsonResponseException);
+                throw googleJsonResponseException;
```
Comment (bergerdenes) [inline cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:245]: you might wrap this into `CloudConnectorException`

**Review by bergerdenes — CHANGES_REQUESTED** [review]

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java (hunk 3 above)
Comment (biharitomi) [reply cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:168]: yes, but we discussed this offline and agreed on to keep the original behaviour in his line of CB. Should I modify it?

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java (hunk 1 above)
Comment (biharitomi) [reply cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/GcpProvisionSetup.java:154]: yes, as the `existingBucket` boolean remains false.

**Review by bergerdenes — APPROVED** [review]

**Pull request merged**